flink-datastream/src/main/java/org/apache/flink/datastream/impl/context/DefaultStateManager.java

```diff
@@ -90,7 +90,19 @@ public <T> Optional<ValueState<T>> getState(ValueStateDeclaration<T> stateDeclar
     }
 
     @Override
-    public <T> Optional<ListState<T>> getState(ListStateDeclaration<T> stateDeclaration)
+    public <T> ValueState<T> getState(ValueStateDeclaration<T> stateDeclaration) throws Exception {
+        Optional<ValueState<T>> stateOptional = getStateOptional(stateDeclaration);
+        if (stateOptional.isEmpty()) {
+            throw new RuntimeException(
+                    "Failed to access the State. You may need to declare it first and ensure that "
+                            + "the context can access the state. For example, retrieving a "
+                            + "keyed state is not allowed in a non-keyed context.");
+        }
```

Comment:
1. Should not use unchecked exceptions.
2. Error message can be deduplicated.
3. Better points to documentation
https://nightlies.apache.org/flink/flink-docs-master/docs/dev/datastream-v2/context_and_state_processing/#the-legitimacy-of-state-declaration-and-access